### DISCUSSION TITLE
Remote and local data fetching

### DISCUSSION THREAD

**Discussion by meixger** [nuxt/nuxt#18404] [Questions]
1. What's recommended way to fetch data from both a remote api and a `~/server/api` in Nuxt 3? 

2. Does Nuxt `$fetch` (unjs/ofetch) have any advantage over Node.js experimental `fetch`?

I struggle with a `Excessive stack depth comparing types` TS error using generics in a custom api client class. [Reproduction repository](https://github.com/meixger/nuxt3-ofetch-test)

![image](https://user-images.githubusercontent.com/600424/213707905-70787c9b-6331-46df-885f-e3fca6467d30.png)

**Comment by oritwoen** [comment]
`$fetch` ([unjs/ofetch](https://github.com/unjs/ofetch)) it has a lot of advantages. First, it uses the native NodeJS environment and native fetch if available, minimizing the bundle. However, if it is not available, such as in the old version of NodeJS or, for example, in another environment (cloudflare, vercel, etc.), it uses the appropriate equivalent.

So it is a package depending on the environment you are using, always works correctly and uses as little JavaScript as possible. When using native fetch on experimental version of NodeJS it will only work on latest versions of NodeJS.

Other advantages are very easy interface, interceptors, auto retry and auto parse response.

As for types, I don't think you're using them properly. I don't think they are even needed in your example because Nuxt has very advanced typing that is automatically created in the background.

Your example can easily be replaced with the code below.

```vue
<script setup lang="ts">
const productUrl = "https://dummyjson.com/products/1";

const { data: product } = await useFetch(productUrl)
const { data: todo } = await useFetch("/api/todo");
</script>

<template>
  <h2>Remote API</h2>
  <p>Product "{{ product?.title }}" </p>
  <h2>Nuxt server-side API</h2>
  <p>Todo "{{ todo?.title }}"</p>
</template>
<style>
body {
  font-family: monospace;
}
</style>
```
